### ISSUE TITLE
perf(anvil): high memory usage when calling `debug_traceTransaction`

### ISSUE THREAD

**Issue by vlad-blana** [foundry-rs/foundry#7275]
### Component

Anvil

### Have you ensured that all of these are up to date?

- [X] Foundry
- [ ] Foundryup

### What version of Foundry are you on?

0.2.0 (551bcb5 2024-02-28T07:42:15.174242400Z)

### What command(s) is the bug in?

.\anvil.exe --fork-url "http://127.0.0.1:8545" --auto-impersonate --steps-tracing

### Operating System

Windows

### Describe the bug

I'm analyzing random erc20 tokens transfer behavior and I found some that seem to deadlock anvil when I request the trace with debug_traceTransaction json rpc call after sending and mining the transcation.

Here is a repro. I'm adding the other transfer tx inputs (to, from, data) soon. I'm extracting them one by one.
https://github.com/vlad-blana/foundry-anvil-lock-repro

It auto starts anvil. You have to copy the anvil.exe in the root folder so the script may find it.

Calling debug_traceTranscation is making anvil reach 14 GB memory usage in some cases.
The transaction receipt and trace_transaction can be retrieved successfully.

No json response is received and sending other json rpc calls after do nothing.

The code is fast on my local node, but I added the fastest and reliable public node around for you to be able to run it and it takes some time to mine the transcation (30s- 1min).

The fun stuff begins only after calling debug_traceTranscation so you have to endure the evm_mine wait before it.
Then the deadlock manifests seconds after the call.
Anvil is active for a few seconds and then it's just inert- no more cpu, memory changes.

**Comment by vlad-blana** [comment]
I've added 10 examples. You have to manually edit the index at line 214. I've mentioned that in the readme too

The last example is the one that makes anvil use 14 GB of ram.

There are a few out of gas/out of stack transaction examples. Maybe they deadlock for a different reason than the the regular status 1 txs.

For most of them anvil reaches 2-3 GB of used memory and it doesn't behave like this for any other erc20 tokens I've tested. I don't know what happens behind the scenes, but it might be normal with how much stuff happens behind the scene in these weird contracts.

I wonder if they also trigger high memory usage on the regular clients, bugging the validators

**Comment by DaniPopes** [comment]
cc @mattss

**Comment by vlad-blana** [comment]
I reviewed the issue and I found out the "lock" was inside ethers.js WebSocketProvider. I checked with JsonRpcProvider and it might finish unless it crashes, but it also using like 3 cpu cores.

Curl or node.js fetch or axios with JSON.parse don't suffer from this issue. All examples generate json text of 150-200MB which is can be handled by js, but ethers fails to.

The high memory usage of anvil is still there. 14 GB for the last example. There are 189164 struct logs so I can't manually figure out what would cause that. I mean it can't be so much state required for only 189164 struct logs, and the other examples have similar structLogs length, but cause anvil to top out at 1-4 GB of ram only

My test results with anvil release 2024-May-2
- they're almost identical on repeated runs and also for the older anvil releases
// 0 - ~ 1GB anvil mem usage - 150351006 json byte size
// 1 -  2727424KB - 241385025 json bytes
// 2 -  3334824KB - 113786048 json bytes
// 3 -  3771620KB - 153234138 json bytes
// 4 -  1257260KB - 125559952 json bytes
// 5 -  2299856KB - 209675456 json bytes
// 6 -  1305192KB - 116816624 json bytes
// 7 -  1228920KB - 121394579 json bytes
// 8 -  1222468KB - 137341881 json bytes
// 9 - 13918872KB - 159530214 json bytes

So the lock wasn't there even when I opened the issue, only the memory issue is what remains. 14 GB ram is dangerous especially if we run other high memory consumption processes on the same machine.

**Comment by grandizzy** [comment]
@vlad-blana would you be able to retest with latest version or provide a version of test drive without the js lock?) thank you!

**Comment by vlad-blana** [comment]
I retested.

In all these instances the issue is ethers js. For some reason it is unable to parse the json it receives past a certain size.
I've used node fetch and axios followed by JSON.parse on the message inside the jsonrpc call and it worked even with the version I opened the issue as well as latest ones. 

The memory usage on the other end is still there. I think these tokens got crazy logic that abuse the hardware. It might be intentional. Most of them are derivations of DistributedDividend tokens. I don't know what's that, but maybe it helps. And there are others too

**Comment by grandizzy** [comment]
thank you! would you be able to share the test drive you used (node fetch and axios), I'd like to give it a try locally and debug / profile.

**Comment by vlad-blana** [comment]
Ok. I'm doing that the next few days

**Comment by vlad-blana** [comment]
I updated [(https://github.com/vlad-blana/foundry-anvil-lock-repro)] to use axios for the debug_traceTransaction call.

The doStuff has 2 more flags to control which one to use: axios http request or ethers http/websocket for debug_traceTransaction only. The other calls are done with ethers since they didn't no cause any problems in my testing.
And I've added more console.log messages to know what's going on while waiting for completion.
The changes from the previous version could be checked with some git viewer or vscode.

I've tested with the latest nightly build and now the max memory usage is 2GB for the last test that used to be 14GB. And the maximum of all is 2.6 GB

In all my tests, the tx mining and receipt used ~40% of the max memory usage. Then the debug_traceTransaction with any method increased the process size to the maximum and after that it cleaned up the memory for debug_traceTransaction - not sure if it was back to the size before it but close to it 500-800MB ram.

I've included measurements and observations at the bottom of anvilLockRepro.js

Ethers is broken with both JsonRpcProvider and WebSocketProvider. With JsonRpcProvider which uses http as the transport protocol it crashes node with some memory error (not out of memory) and WebSocketProvider may or may not return the trace but in all instances it did not allow anvil to close - and after I added additional log messages it didn't return the trace so that the process kill line was not reached to see.
Anyway. I don't use ethers for debug_traceTransaction in a while because of these bugs and others it has when doing concurrent calls for example, so I don't care about this freeze issue and probably others don't either.

What remains is, if it makes sense to you, to check if the 2GB+ memory usage is within parameters, if anvil allocates and deallocates memory properly and efficiently (no redundant data here and there)
